plugins/device/cmd/nvidia/device.go

```diff
@@ -0,0 +1,260 @@
+type Config struct {
+	// TODO(oleksii-shyman):: this has to be a slice, not csv
+	// but I can't figure out how to specify a slice of strings
+	BlacklistedGPUIds string `codec:"blacklisted_gpu_ids"`
+	FingerprintPeriod string `codec:"fingerprint_period"`
+}
+
+type NvidiaDevice struct {
+	logger log.Logger
+
+	// BlacklistedGPUIds is a list of UUIDs that would not be exposed to nomad
+	BlacklistedGPUIds []string
+	// fingerprintPeriod is how often we should call nvml to get list of devices
+	fingerprintPeriod time.Duration
+
+	// devices is the set of detected eligible devices and maps whether they are healthy
+	devices    map[string]bool
+	deviceLock sync.RWMutex
+}
+
+// NewExampleDevice returns a new example device plugin.
+func NewNvidiaDevice(log log.Logger) *NvidiaDevice {
+	return &NvidiaDevice{
+		logger:  log.Named(pluginName),
+		devices: make(map[string]bool),
+	}
+}
+
+// PluginInfo returns information describing the plugin.
+func (d *NvidiaDevice) PluginInfo() (*base.PluginInfoResponse, error) {
+	// TODO(oleksii-shyman)::how this plugin api versions would be used?
+	return pluginInfo, nil
+}
+
+// ConfigSchema returns the plugins configuration schema.
+func (d *NvidiaDevice) ConfigSchema() (*hclspec.Spec, error) {
+	return configSpec, nil
+}
+
+// SetConfig is used to set the configuration of the plugin.
+func (d *NvidiaDevice) SetConfig(data []byte) error {
+	var config Config
+	if err := base.MsgPackDecode(data, &config); err != nil {
+		return err
+	}
+
+	// TODO(oleksii-shyman):: this has to be a slice, not csv
+	// but I can't figure out how to specify a slice of strings
+	d.BlacklistedGPUIds = strings.Split(config.BlacklistedGPUIds, ",")
+
+	period, err := time.ParseDuration(config.FingerprintPeriod)
+	if err != nil {
+		return fmt.Errorf("failed to parse list period %q: %v", config.FingerprintPeriod, err)
+	}
+	d.fingerprintPeriod = period
+
+	return nil
+}
+
+// Fingerprint streams detected devices. If device changes are detected or the
+// devices health changes, messages will be emitted.
+func (d *NvidiaDevice) Fingerprint(ctx context.Context) (<-chan *device.FingerprintResponse, error) {
+	outCh := make(chan *device.FingerprintResponse)
+	go d.fingerprint(ctx, outCh)
+	return outCh, nil
+}
+
+// composeFingerprintDeviceList filters blacklisted devices and transforms nvml driver output to []*device.Device type
+func composeFingerprintDeviceList(deviceData []*nvml.FingerprintDeviceData, blacklistedGPUIds []string) []*device.Device {
+
+	blacklistedMap := make(map[string]struct{})
+	for _, blacklistedGPUId := range blacklistedGPUIds {
+		blacklistedMap[blacklistedGPUId] = struct{}{}
+	}
+
+	var result []*device.Device
+	for _, fingerprintDevice := range deviceData {
+		if _, blacklisted := blacklistedMap[fingerprintDevice.UUID]; !blacklisted {
+			result = append(result, &device.Device{
+				ID: fingerprintDevice.UUID,
+				// all nonblacklisted devices are "healthy", because nvidia
```

Comment:
Might want to check if health of a card depends on the temperature of the device.